[rum] added beforesend context docs (#10829)
DataDog/documentation: content/en/real_user_monitoring/browser/advanced_configuration.md

      text: 'Tracking custom user actions'
---

## Enrich and control RUM data

> The RUM SDK automatically captures most RUM events and populates their main attributes. The `beforeSend` callback function gives you access to every event collected by the RUM SDK before they get sent to Datadog. Intercepting the RUM events allows you to:

Reviewer: ```suggestion
The RUM SDK captures RUM events and populates their main attributes. The `beforeSend` callback function gives you access to every event collected by the RUM SDK before they are sent to Datadog. Intercepting the RUM events allows you to:
```

> * Enrich your RUM events with extra context attributes

Reviewer: ```suggestion
* Enrich your RUM events with additional context attributes
```

> * Modify your RUM events to modify their content, or redact sensitive sequences

Reviewer 2: ```suggestion
* Redact sensitive sequences of your RUM events
```
We still can't modify the content of the event, only some allowed fields

Author: Let me add a note "(see list of editable attributes[link])"

* Discard selected RUM events

```javascript
function beforeSend(event, context)
```

> Starting with [version 2.13.0][3], `beforeSend` takes 2 arguments: the `event` generated by the RUM SDK, and the `context` that triggered the creation of the RUM event. Here is the list of potential `context` values:

Reviewer: ```suggestion
Starting with [version 2.13.0][3], `beforeSend` takes two arguments: the `event` generated by the RUM SDK, and the `context` that triggered the creation of the RUM event. The potential `context` values are:
```

| RUM event type   | Context                   |
|------------------|---------------------------|
| View             | [Location][4]                  |
| Action           | [Event][5]                     |
| Resource (XHR)   | [XMLHttpRequest][6]            |

> | Resource (Fetch) | [Request][7] and [Response][8]      |

Reviewer 2: we also have the corresponding PerformanceResourceTiming when available

Reviewer 3: Yes, I think this table could be improved with exact properties and types, at least for JS developers who won't have type hints: https://github.com/DataDog/browser-sdk/blob/main/packages/rum-core/src/rawRumEvent.types.ts#L235-L260

Author: I tried to balance between exhaustiveness and making it understandable for the average reader. I also can link to the file but then if we change anything the link is broken... I'm going to add `PerformanceResourceTiming` and that will do the trick. What do you think?

Reviewer 2: It sounds ok-ish if we plan on having a more complete guide quickly, otherwise I'd be in favor of having something more exhaustive first and simplify when we will have a dedicated guide.

Author: There's just `Error` missing now then right? What's the difference between `RequestInit` and `RequestInput`?

| Resource (Other) | [PerformanceResourceTiming][9] |
| Error            | [Error][10]                     |
| Long Task        | [PerformanceLongTaskTiming][11] |

### Enrich RUM events

> On top of attributes added with the [global context API](#global-context), you can add extra context attributes to the event. For example, tag your RUM resource events with data extracted from a fetch response object:

Reviewer: ```suggestion
Along with attributes added with the [global context API](#global-context), you can add more context attributes to the event. For example, tag your RUM resource events with data extracted from a fetch response object:
```

datadogRum.init({
    ...,
    beforeSend: (event, context) => {
        // collect a RUM resource's response headers
        if (event.resource && event.resource.type === 'fetch') {
            event.context = {...event.context, responseHeaders: context.response.headers}
        }
    },
    ...
});
```

{{% /tab %}}
{{% tab "CDN async" %}}
```javascript
DD_RUM.onReady(function() {
    DD_RUM.init({
        ...,
        beforeSend: (event, context) => {
            // collect a RUM resource's response headers
            if (event.resource && event.resource.type === 'fetch') {
                event.context = {...event.context, responseHeaders: context.response.headers}
            }
        },
        ...
    })
})
```
{{% /tab %}}
{{% tab "CDN sync" %}}

```javascript
window.DD_RUM &&
    window.DD_RUM.init({
        ...,
        beforeSend: (event, context) => {
            // collect a RUM resource's response headers
            if (event.resource && event.resource.type === 'fetch') {
                event.context = {...event.context, responseHeaders: context.response.headers}
            }
        },
        ...
    });
```

{{% /tab %}}
{{< /tabs >}}

> **Note**: The RUM SDK will ignore attributes added outside of `event.context`. It will also ignore modifications made to a RUM view event context.

Reviewer: ```suggestion
**Note**: The RUM SDK ignores:
- Attributes added outside of `event.context`.
- Modifications made to a RUM view event context.
```

|   `error.stack `        |   String  |   The stack trace or complementary information about the error.                                     |
|   `error.resource.url`  |   String  |   The resource URL that triggered the error.                                                        |
|   `resource.url`        |   String  |   The resource URL.                                                                                 |
|   `context`        |   Object  |   Attributes added via the [global context API](#global-context) or when generating events manually (for example, `addError` and `addAction`). RUM view events `context` is read-only.                                                                                 |

> **Note**: The RUM SDK will ignore modifications made to event properties not listed above. Find out about all event properties on the [Browser SDK repository][1].

Reviewer: ```suggestion
**Note**: The RUM SDK ignores modifications made to event properties not listed above. Find out about all event properties on the [Browser SDK repository][1].
```
